packages/atomic/src/utils/init-queue.ts

```diff
@@ -0,0 +1,68 @@
+import {InitializeEventHandler} from './initialization-lit-stencil-common-utils';
+
+export type InitializeEvent = CustomEvent<InitializeEventHandler>;
+
+declare global {
+  interface Window {
+    initQueueNamespace: {
+      eventQueueMap: Map<Element, QueuedEvent[]>;
+      parentReadyMap: Map<Element, boolean>;
+    };
+  }
+}
+
+type QueuedEvent = {
+  event: InitializeEvent;
+  element: Element;
+};
+
+function getWindow() {
+  return window;
+}
+
+function getInitQueueNamespace() {
+  if (!getWindow().initQueueNamespace) {
+    getWindow().initQueueNamespace = {
+      eventQueueMap: new Map(),
+      parentReadyMap: new Map(),
+    };
+  }
+  return getWindow().initQueueNamespace;
+}
+
+function getEventQueueMap(): Map<Element, QueuedEvent[]> {
+  return getInitQueueNamespace().eventQueueMap;
+}
+
+function getParentReadyMap(): Map<Element, boolean> {
+  return getInitQueueNamespace().parentReadyMap;
+}
+
+export function markParentAsReady(parent: Element) {
+  const parentReadyMap = getParentReadyMap();
+  parentReadyMap.set(parent, true);
+  const eventQueueMap = getEventQueueMap();
+  const eventQueue = eventQueueMap.get(parent) || [];
+  while (eventQueue.length > 0) {
```

Comment:
This is inefficient because `Array.prototype.shift` reindexes the array on every call. Can't we use `.pop()` instead since it is an `O(1)` operation?

If the dispatch order is really important, then we should have a test for that.

Reply:
Good point. IIRC, order matters (i.e. we need a FIFO structure, i.e. queue. Stack doesn't cut it)
So here we have a O(n^2). oops.
Assuming the queue is immutable once we start reading it, we should be able to make due with O(n) by reversing the array before reading it with pop.

Reply:
If we want to go further, we could go O(1) but we'd need to craft a LinkedList, but that's fricking overkill when we consider that this is only there till we get rid of Stencil.